### DISCUSSION TITLE
Load a Device tree overlay at boot time with Uboot and pass it to Kernel

### DISCUSSION THREAD

**Discussion by Alb7954** [polarfire-soc/polarfire-soc-documentation#457] [Linux build systems for PolarFire SoC] [answered]
Hello,

when compiling one of the reference image that includes the "dt-overlay-mchp" package, the 4 dtso files available in linux4microchip/dt-overlay-mchp repository are proper compiled and the dtbo files are placed in the /boot target folder, as expected. The FIT image is also created with all the overlays included.

In general, they could be loaded at boot time using bootm command in Uboot, but it's not clear to me how to do it with the provided Uboot Env infrastructure. I see a "design_overlays" procedure that depends on "no_of_overlays" variable in Uboot Env, but I do not understand how to use it to select a specific overlay from one of the 4 included in the fit image.
Is there any simple example to look at in order to understand how to load one or more of the overlays and pass it to the Kernel ?

Thanks

**Comment by vfalanis** [comment, marked as answer]
Hello @Alb7954 ,

The device tree overlays provided as part of the dt-overlays-mchp repository can be loaded by U-boot using the bootm command. This is particularly useful for adding  Click boards or Raspberry Pi HATs without needing to recompile the entire device tree blob.

We have a few examples of this in the BSP, specifically for loading the qspi nand and nor overlays. Please refer to [this](https://github.com/polarfire-soc/meta-polarfire-soc-yocto-bsp/blob/master/meta-polarfire-soc-bsp/recipes-bsp/u-boot/files/icicle-kit-es-nor/boot.cmd#L6) U-boot source script. 

As shown in the link provided you could edit the boot.cmd source script in meta-polarfire-soc-bsp/recipes-bsp/u-boot/files/<MACHINE>boot.cmd and edit the bootm start command to append an overlay. For example:

```
bootm start ${scriptaddr}#conf-microchip_mpfs-icicle-kit.dtb#conf-mpfs_icicle_pmod_sf3.dtbo
```

The command above is an example of how you can use the bootm command in U-Boot to load a specific device tree overlay along with the primary device tree blob. The `conf-microchip_mpfs-icicle-kit.dtb` is the primary device tree blob for the Microchip MPFS Icicle Kit. The `conf-mpfs_icicle_pmod_sf3.dtbo` is the device tree overlay configuration for the PMOD SF3, which extends the main device tree to support this specific click board. The hash symbol is used to separate the main device tree blob (.dtb) from the overlays (.dtbo).

It is worth mentioning that the overlay must be part of the fitImage to use the command above. In Yocto images, the FitImage is located in the first partition of the MMC/SD.

It is also possible to apply overlays located in the /boot directory at runtime from the kernel. For this method, please see [this](https://github.com/polarfire-soc/polarfire-soc-linux-examples/tree/master/dt-overlays) document.

On the other hand, the design_overlays command in U-boot is meant for a different purpose. This functionality is used to load device tree overlays from the internal SPI flash. This method is used for loading overlays related to fabric content at boot.

Hope this clarifies things.

**Reply by Alb7954** [reply]
Thank you. 
I'll check what you suggested.
